File: vscode-tests/vscode-onig/src/lib/onig-string.ts
import { IOnigBinding, Pointer } from "./types";
import UtfString from "./utf-string";

const onigBinding: IOnigBinding | null = null;

export class OnigString {
  private static LAST_ID = 0;
  private static _sharedPtr: Pointer = 0; // a pointer to a string of 10000 bytes
  private static _sharedPtrInUse = false;

  public readonly id = (++OnigString.LAST_ID);
  private readonly _onigBinding: IOnigBinding;
  public readonly content: string;
  public readonly utf16Length: number;
  public readonly utf8Length: number;
  public readonly utf16OffsetToUtf8: Uint32Array | null;
  public readonly utf8OffsetToUtf16: Uint32Array | null;
  public readonly ptr: Pointer;

  constructor(str: string) {
    if (!onigBinding) {
      throw new Error(`Must invoke loadWASM first.`);
    }
    this._onigBinding = onigBinding;
    this.content = str;
    const utfString = new UtfString(str);
    this.utf16Length = utfString.utf16Length;
    this.utf8Length = utfString.utf8Length;
    this.utf16OffsetToUtf8 = utfString.utf16OffsetToUtf8;
    this.utf8OffsetToUtf16 = utfString.utf8OffsetToUtf16;

    if (this.utf8Length < 10000 && !OnigString._sharedPtrInUse) {
      if (!OnigString._sharedPtr) {
        OnigString._sharedPtr = onigBinding._malloc(10000);
      }
      OnigString._sharedPtrInUse = true;
      onigBinding.HEAPU8.set(utfString.utf8Value, OnigString._sharedPtr);
      this.ptr = OnigString._sharedPtr;
    } else {
      this.ptr = utfString.createString(onigBinding);
    }
  }

  public convertUtf8OffsetToUtf16(utf8Offset: number): number {
    if (this.utf8OffsetToUtf16) {
      if (utf8Offset < 0) {
        return 0;
      }
      if (utf8Offset > this.utf8Length) {
        return this.utf16Length;
      }
      return this.utf8OffsetToUtf16[utf8Offset];
    }
    return utf8Offset;
  }

  public convertUtf16OffsetToUtf8(utf16Offset: number): number {
    if (this.utf16OffsetToUtf8) {
      if (utf16Offset < 0) {
        return 0;
      }
      if (utf16Offset > this.utf16Length) {
        return this.utf8Length;
      }
      return this.utf16OffsetToUtf8[utf16Offset];
    }
    return utf16Offset;
  }

  public dispose(): void {
    if (this.ptr === OnigString._sharedPtr) {
      OnigString._sharedPtrInUse = false;
    } else {
      this._onigBinding._free(this.ptr);
    }
  }
}

File: scie-scanner/src/scanner/onig_string.rs
use crate::scanner::utf_string::UtfString;

#[derive(Debug, Clone)]
pub struct OnigString<'a> {
    pub id: i32,
    pub content: &'a str,
    pub utf16length: i32,
    pub utf8length: i32,
    pub utf16offset_to_utf8: Vec<u32>,
    pub utf8offset_to_utf16: Vec<u32>,
}

impl<'a> OnigString<'a> {
    pub fn new(str: &str, id: i32) -> OnigString {
        let utf_string = UtfString::new(str);
        OnigString {
            id,
            content: str,
            utf16length: utf_string.utf16length,
            utf8length: utf_string.utf8length,
            utf16offset_to_utf8: utf_string.utf16offset_to_utf8,
            utf8offset_to_utf16: utf_string.utf8offset_to_utf16,
        }
    }

    pub fn convert_utf8offset_to_utf16(&self, utf8offset: i32) -> i32 {
        if self.utf8offset_to_utf16.len() > 0 {
            if utf8offset < 0 {
                return 0;
            }
            if utf8offset > self.utf8length {
                return self.utf16length;
            }
            return self.utf8offset_to_utf16[utf8offset as usize] as i32;
        }
        return utf8offset;
    }

    pub fn convert_utf16offset_to_utf8(&self, utf16offset: i32) -> i32 {
        if self.utf16offset_to_utf8.len() > 0 {
            if utf16offset < 0 {
                return 0;
            }
            if utf16offset > self.utf16length {
                return self.utf8length;
            }

            return self.utf16offset_to_utf8[utf16offset as usize] as i32;
        }

        return utf16offset;
    }
}

#[cfg(test)]
mod tests {
    use crate::scanner::onig_string::OnigString;

    #[test]
    fn should_handle_offset() {
        let onig_string = OnigString::new("a💻bYX", 1);
        let x = onig_string.convert_utf8offset_to_utf16(2);
        assert_eq!(1, x);
    }
}
